builder/packages_image.go

```diff
@@ -39,86 +40,135 @@ func NewPackagesImageBuilder(repository, compiledPackagesPath, targetPath, fissi
+	root   string      // The base directory on disk where the walking started
+	prefix string      // The prefix in the tar file the names should have
+}
 
-	succeeded := false
-	buildDir, err := ioutil.TempDir(p.targetPath, "role-packages")
+func (w *tarWalker) walk(path string, info os.FileInfo, err error) error {
+	if err != nil {
+		return err
+	}
+	header, err := tar.FileInfoHeader(info, "")
 	if err != nil {
-		return "", err
+		return err
 	}
-	defer func() {
-		if !succeeded {
-			os.RemoveAll(buildDir)
+	if (info.Mode() & os.ModeSymlink) != 0 {
+		linkname, err := os.Readlink(path)
+		if err != nil {
+			return err
 		}
-	}()
-	rootDir := filepath.Join(buildDir, "root")
-
-	// Generate dockerfile
-	dockerfile, err := os.Create(filepath.Join(buildDir, "Dockerfile"))
+		header.Linkname = linkname
+	}
+	relPath, err := filepath.Rel(w.root, path)
 	if err != nil {
-		return "", err
+		return err
+	}
+	header.Name = filepath.Join(w.prefix, relPath)
+	if err := w.stream.WriteHeader(header); err != nil {
+		return err
 	}
-	defer dockerfile.Close()
-	baseImageName := GetBaseImageName(p.repository, p.fissileVersion)
-	if err := p.generateDockerfile(baseImageName, dockerfile); err != nil {
-		return "", err
+	if info.Mode().IsRegular() {
+		file, err := os.Open(path)
+		if err != nil {
+			return err
+		}
+		defer file.Close()
+		if _, err = io.CopyN(w.stream, file, info.Size()); err != nil {
+			return err
+		}
 	}
+	return nil
+}
 
-	// Generate configuration
-	p.ui.Println("Generating configuration JSON specs ...")
-	configStore := configstore.NewConfigStoreBuilder(
-		configstore.JSONProvider,
-		lightManifestPath,
-		darkManifestPath,
-		filepath.Join(rootDir, "opt/hcf/specs"),
-	)
-
-	if err := configStore.WriteBaseConfig(roleManifest); err != nil {
-		return "", fmt.Errorf("Error writing base config: %s", err.Error())
+// CreatePackagesDockerStream generates a tar stream containing the docker context to build the packages layer image with
+func (p *PackagesImageBuilder) CreatePackagesDockerStream(roleManifest *model.RoleManifest, lightManifestPath, darkManifestPath string) (io.ReadCloser, <-chan error, error) {
+	if len(roleManifest.Roles) == 0 {
+		return nil, nil, fmt.Errorf("No roles to build")
 	}
 
-	// Copy packages
-	p.ui.Println("Copying compiled packages...")
-	copiedFingerprints := make(map[string]bool)
-	for _, role := range roleManifest.Roles {
-		for _, job := range role.Jobs {
-			for _, pkg := range job.Packages {
-				if _, ok := copiedFingerprints[pkg.Fingerprint]; ok {
-					// Package has already been copied (possibly due to a different role)
-					continue
-				}
+	pipeReader, pipeWriter := io.Pipe()
+	errors := make(chan error)
 
-				compiledDir := pkg.GetPackageCompiledDir(p.compiledPackagesPath)
-				if err := util.ValidatePath(compiledDir, true, fmt.Sprintf("compiled dir for package %s", pkg.Name)); err != nil {
-					return "", err
-				}
+	go func() {
+		defer close(errors)
+		defer pipeWriter.Close()
```

Comment:
No, because that gets passed to the docker client. They can close it when they're done reading (when it hits EOF).  In fact, I _shouldn't_ close it for them, otherwise they'd get a short read.